src/native/managed/cdacreader/Microsoft.Diagnostics.DataContractReader.Contracts/Contracts/ExecutionManager/Helpers/NibbleMapHelpers.cs

```diff
@@ -0,0 +1,147 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Numerics;
+using System.Diagnostics;
+using System;
+
+namespace Microsoft.Diagnostics.DataContractReader.ExecutionManagerHelpers;
+
+internal static class NibbleMapHelpers
+{
+    // we will partition the address space into buckets of this many bytes.
+    // There is at most one code block header per bucket.
+    // Normally we would then need 5 bits (Log2(BytesPerBucket))to find the exact start address,
+    // but because code headers are aligned, we can store the offset in a 4-bit nibble instead and shift appropriately to compute
+    // the effective address
+    internal const ulong BytesPerBucket = 8 * MapUnit.SizeInBytes;
+
+    // We load the map contents as 32-bit integers, which contains 8 4-bit nibbles.
+    // The algorithm will focus on each nibble in a map unit before moving on to the previous map unit
+    internal readonly struct MapUnit
+    {
+        public const int SizeInBytes = sizeof(uint);
+        public const ulong SizeInNibbles = 2 * SizeInBytes;
+        public readonly uint Value;
+
+        public MapUnit(uint value) => Value = value;
+
+        public override string ToString() => $"0x{Value:x}";
+
+        // Shift the next nibble into the least significant position.
+        public MapUnit ShiftNextNibble => new MapUnit(Value >>> 4);
+
+        public const uint NibbleMask = 0x0Fu;
+        internal Nibble Nibble => new((int)(Value & NibbleMask));
+
+        public bool IsZero => Value == 0;
```

Comment:
Should this be `IsEmpty` to match `Nibble.IsEmpty`?

Reply:
Agreed